include/ibamr/MassIntegrator.h

```diff
@@ -0,0 +1,301 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2018 - 2020 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBAMR_MassIntegrator
+#define included_IBAMR_MassIntegrator
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include "ibamr/StaggeredStokesPhysicalBoundaryHelper.h"
+#include "ibamr/ibamr_enums.h"
+
+#include "ibtk/CartGridFunction.h"
+#include "ibtk/HierarchyGhostCellInterpolation.h"
+#include "ibtk/HierarchyMathOps.h"
+#include "ibtk/ibtk_utilities.h"
+
+#include "CoarseFineBoundary.h"
+#include "HierarchySideDataOpsReal.h"
+#include "IntVector.h"
+#include "PatchHierarchy.h"
+#include "SideVariable.h"
+#include "tbox/Database.h"
+#include "tbox/DescribedClass.h"
+#include "tbox/Pointer.h"
+
+#include <array>
+#include <string>
+#include <vector>
+
+namespace SAMRAI
+{
+namespace solv
+{
+template <int DIM, class TYPE>
+class SAMRAIVectorReal;
+template <int DIM>
+class RobinBcCoefStrategy;
+} // namespace solv
+namespace hier
+{
+template <int DIM>
+class BasePatchHierarchy;
+} // namespace hier
+} // namespace SAMRAI
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class MassIntegrator is an abstract class which integrates
+ * the density field
+ *
+ *  \f$ \frac{\partial \rho}{\partial t} + \nabla \cdot (\rho u) = S(x,t) \f$
+ *
+ * and computes the conservative form of the convective operator
```

Comment:
I think one of the points is that you need to make these all consistent to get good results --- maybe could add that in the comments (otherwise it seems a little odd that these other terms are being computed in a solver for the mass).